keps/sig-auth/20190116-service-account-external-signing.md

```diff
@@ -0,0 +1,152 @@
+  - TBD
+editor: TBD
+creation-date: 2019-01-16
+last-updated: 2019-01-16
+status: provisional
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Support external signing of service account keys
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Preserve existing behavior](#preserve-existing-behavior)
+  * [Updates to API server token generation](#updates-to-api-server-token-generation)
+  * [New API](#new-api)
+  * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+The Kubernetes API server has always read service account keys from disk as the process starts, and kept them in memory for the duraiton of the server's lifetime. As the API server can now verify and issue projected volume tokens, it would be advantageous to support external signing and verifying of token data over an API, as well as reading public keys from an API.
+
+## Motivation
+
+For operators who want to regularly rotate the signing and verifying keys for projected volume tokens, the Kubernetes API server must be restarted in order to use a new key. To facilitate easy key rotation, this KEP includes an proposal for a grpc API to support out of process signing and listing of signing keys.
+
+### Goals
+
+- Support for out-of-process JWT signing
+- Support for listing public verifying keys
+- Preserve existing behavior and performance for keys not read over a socket
+
+### Non-Goals
+
+- Reading TLS serving certificates and key from a socket or reloading of the API server with new cert and key
+- Reading any other certificates from a file
+
+## Proposal
+
+### Preserve existing behavior
+
+The API server flags `--service-account-key-file` and `--service-account-signing-key-file` will continue be used for reading from files. The API server will detect if the provided files are regular files to be read once at process startup, or unix sockets that should be read on each token request using a new API. Legacy service account verification will skip any socket files.
+
+### Updates to API server token generation
+
+As of Kubernetes v1.13.2, the API server uses the functions `JWTTokenGenerator` and `JWTTokenAuthenticator`. New types that implement the `TokenGenerator` interface and support token validation will be added to `k8s.io/kubernetes/pkg/serviceaccount/`.
+
+### New API
+
+I'm proposing creating a new versioned grpc API under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`
+
+```proto
+syntax = "proto3";
+
+package v1beta1;
+
+service JWTKeyService {
```

Comment:
I understand this KEP is for JWT signing only, but the proposed API looks applicable in other cases (like TLS client/server keys).
Can this grpc service be renamed to `KeyService` or something similar for future use? The KEP should still only use it for JWTs.

Reply:
Sure. I'll update this